### PR TITLE
[type: refactor] optimize & fix uri with invalid char.

### PR DESCRIPTION
**Pull request by qicz**
fix `status=[abc%2Cedf]`query  `Invalid character '[' for QUERY_PARAM in status=[abc%2Cedf]` bug

### CODE REVIEW

File: shenyu-plugin/shenyu-plugin-uri/src/main/java/org/apache/shenyu/plugin/uri/URIPlugin.java [hunk 1]
```diff
@@ -42,27 +42,18 @@
...
-            path = String.join("?", path, uri.getRawQuery());
-            realURI = UriComponentsBuilder.fromHttpUrl(path).build(true).toUri();
-        } else {
-            if (StringUtils.isNotEmpty(uri.getQuery())) {
-                path = String.join("?", path, uri.getQuery());
-            }
-            realURI = UriComponentsBuilder.fromHttpUrl(path).build(false).toUri();
-        }
-        exchange.getAttributes().put(Constants.HTTP_URI, realURI);
+        URI realUri = URI.create(path + "?" + RequestQueryCodecUtil.getCodecQuery(exchange));
```
Comment (yu199195) [inline shenyu-plugin/shenyu-plugin-uri/src/main/java/org/apache/shenyu/plugin/uri/URIPlugin.java:55]: why modify this， only used  UriUtils.encode(item, StandardCharsets.UTF_8))？

**Review by yu199195 — CHANGES_REQUESTED** [review]
not merge

File: shenyu-plugin/shenyu-plugin-uri/src/main/java/org/apache/shenyu/plugin/uri/URIPlugin.java (hunk 1 above)
Comment (qicz) [reply shenyu-plugin/shenyu-plugin-uri/src/main/java/org/apache/shenyu/plugin/uri/URIPlugin.java:55]: some query like `status=[abc%2Cedf]` will catch the `Invalid character '[' for QUERY_PARAM in status=[abc%2Cedf]`

Comment (qicz) [reply shenyu-plugin/shenyu-plugin-uri/src/main/java/org/apache/shenyu/plugin/uri/URIPlugin.java:55]: I will remove the `encode` transfer of the original data to upstream.

**Comment by qicz** [comment]
> not merge

it is a bug.

some query like status=[abc%2Cedf] will catch the Invalid character '[' for QUERY_PARAM in status=[abc%2Cedf]

**Comment by yu199195** [comment]
If modified this way, all of them are used url encoder. But for some normal ones, it is not necessary.

> some query like status=[abc%2Cedf] will catch the Invalid character '[' for QUERY_PARAM in status=[abc%2Cedf]

**Comment by qicz** [comment]
> If modified this way, all of them are used url encoder. But for some normal ones, it is not necessary.
> 
> > some query like status=[abc%2Cedf] will catch the Invalid character '[' for QUERY_PARAM in status=[abc%2Cedf]

A framework like `spring` will decode automatically.

**Comment by qicz** [comment]
@yu199195 I have refactored this and transferred the original data to upstream, just encode ' ' (the empty string).

**Comment by qicz** [comment]
@yu199195 I has been tested, support special characters, support partial encoding, transparent transmission of business data, except for Spaces, other is unencoded raw data

**Review by dengliming — COMMENTED** [review]
We need integration testing for these changes.

**Comment by qicz** [comment]
> We need integration testing for these changes.

I am trying to test but the integration uses the ok HTTP client, and it is different with WebClient.

there are many example HTTP run results
image
image
image
image
image

**Review by yu199195 — APPROVED** [review]

**Pull request merged**